syntax = "proto3";

package modules;

import "google/protobuf/empty.proto";

service ModulesService {
    rpc Register(RegisterRequest) returns (RegisterResponse);
    rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
    rpc Deregister(DeregisterRequest) returns (google.protobuf.Empty);
    rpc ListServices(ListServicesRequest) returns (ListServicesResponse);
    rpc GetService(GetServiceRequest) returns (GetServiceResponse);
}

message RegisterRequest {
    string service_name = 1;
    string version = 2;
    string admin_topic = 3;
}

message RegisterResponse {
    string instance_id = 1;
    int32 ttl_seconds = 2;
}

message HeartbeatRequest {
    string instance_id = 1;
}

message HeartbeatResponse {
    bool ok = 1;
    int32 ttl_seconds = 2;
}

message DeregisterRequest {
    string instance_id = 1;
}

message ListServicesRequest {}

message ListServicesResponse {
    repeated ServiceInfo services = 1;
}

message ServiceInfo {
    int32 id = 1;
    string service_name = 2;
    string version = 3;
    string admin_topic = 4;
    int32 ttl_seconds = 5;
    string status = 6;
    string created_at = 7;
    string updated_at = 8;
}

message GetServiceRequest {
    string service_name = 1;
}

message GetServiceResponse {
    ServiceInfo service = 1;
}
